### PR TITLE
Accept all text-like input fields in Textfield-related keywords

### LINKED ISSUE #546: HTML5 specialized text fields not recognized as text fields
HTML5 introduced several new input types, several of which are specialized text inputs including email,  number, url, search, and tel. These input types are not recognized by `Page Should Contain Textfield`, `Page Should Not Contain Textfield`, `Textfield Should Contain`,  or `Textfield Value Should Be`.
The cause is that these keywords use the constraint 'text field' and ElementFinder will filter out any input elements that do not have type 'text'. Since these inputs are still text inputs, they should be treated as such. The number type may render a spinbox control in some browsers, but still behaves like a text-type input.

Here is a test that should pass, but fails with:
Value of text field 'name=email' should have been 'selenium@xxx.yyy' but was 'None'

```
Open Browser    http://www.the-art-of-web.com/html/html5-form-validation/   gc
Input Text  name=email  selenium@xxx.yyy
Textfield Value Should Be   name=email  selenium@xxx.yyy
```

With minor modifications to elementfinder, I was able to get the above to pass.

This is being entered as a result of this: http://stackoverflow.com/questions/34169390/how-to-find-value-of-email-field-using-selenium2library

### PR DESCRIPTION
**Pull request by leods92**
This fixes #546

### CODE REVIEW

File: src/Selenium2Library/locators/elementfinder.py [hunk 1]
```diff
@@ -192,6 +194,8 @@ def _element_matches(self, element, tag, constraints):
         if not element.tag_name.lower() == tag:
             return False
         for name in constraints:
+            if type(constraints[name]) is list and element.get_attribute(name) not in constraints[name]:
+                return False
```
Comment (pekkaklarck) [inline src/Selenium2Library/locators/elementfinder.py]: - Why is this change needed?
- Should use `usinstance`, not `tyoe`

**Comment by pekkaklarck** [comment]
I'm +1 for this enhancement but tests are needed and this should be mentioned in docs too.

**Comment by aaltat** [comment]
And raise issue also from the problem, so that release notes are properly created.

**Comment by pekkaklarck** [comment]
@aaltat There's already #546 about this.

**Comment by aaltat** [comment]
Ah, I did not remember that one, thanks Pekka.  

In any case, the changes needs to be tested. In in unit and/or acceptance level. Also the change breaks the existing unit test which is not acceptable. 

For the code side, I do not have other comments than Pekka did made. Could you fix the testing side related problems and lets see how it continues from there.

File: src/Selenium2Library/locators/elementfinder.py (hunk 1 above)
Comment (leods92) [reply src/Selenium2Library/locators/elementfinder.py]: > Why is this change needed?

Currently constraints only accept 1 value per element attribute and the change involves the creation of a list of values that should be considered acceptable.

> Should use usinstance, not tyoe

Will do the change. Never coded in Python so I really don't know what's best. :) Thanks for the review! 👍

**Comment by leods92** [comment]
@aaltat 

> In any case, the changes needs to be tested. In in unit and/or acceptance level. Also the change breaks the existing unit test which is not acceptable.

> Could you fix the testing side related problems and lets see how it continues from there.

For sure! When I opened the pull request I was busy at work. I was hoping somebody experienced would take it over. In any case, I have some time know and I'm gonna take the time to learn a bit about testing in Python. If I cannot successfully finish it, I'll let you guys know.

**Comment by leods92** [comment]
@pekkaklarck 

Could you please have a second look?
I tried to make tests diversified by testing 2 additional input types: `email` and `url`.
I believe they are the most used text-like input types after `text`.

I hope that's enough, if it's not, please let me know how I can improve. :)

Thanks!

**Comment by pekkaklarck** [comment]
I'm on vacation without my laptop next week (and currently trying to get everything packed). I can take a quick look at the code now, but hopefully @aaltat has time for a full review.

File: src/Selenium2Library/locators/elementfinder.py [hunk 2]
```diff
@@ -151,6 +151,17 @@ def _find_by_default(self, browser, criteria, tag, constraints):
...
+    def _get_xpath_constraints(self, constraints):
+        xpath_constraints = []
+        for name in constraints:
+            value = constraints[name]
+            if isinstance(value, list):
+                contraint = "@%s[. = '%s']" % (name, "' or . = '".join(value))
+            else:
+                contraint = "@%s='%s'" % (name, value)
+            xpath_constraints.append(contraint)
+        return xpath_constraints
```
Comment (pekkaklarck) [inline src/Selenium2Library/locators/elementfinder.py]: This seems to be totally fine, but if there would be `_get_xpath_constraint`, looping could be handled by a list comprehension similarly as earlier:

```python
xpath_constraints = [self._get_xpath_constraint(name, value)
                     for name, value in constraints.items()]
```

File: src/Selenium2Library/locators/elementfinder.py [hunk 3]
```diff
@@ -192,7 +205,10 @@ def _element_matches(self, element, tag, constraints):
         if not element.tag_name.lower() == tag:
             return False
         for name in constraints:
-            if not element.get_attribute(name) == constraints[name]:
+            if isinstance(constraints[name], list):
+                if element.get_attribute(name) not in constraints[name]:
+                    return False
+            elif not element.get_attribute(name) == constraints[name]:
```
Comment (pekkaklarck) [inline src/Selenium2Library/locators/elementfinder.py]: Using `!=` would be more clear than `not` and `==`.

Comment (leods92) [reply src/Selenium2Library/locators/elementfinder.py]: I agree but I thought you Python guys did that (based on the line removed above). Change made! :)

File: src/Selenium2Library/locators/elementfinder.py (hunk 2 above)
Comment (leods92) [reply src/Selenium2Library/locators/elementfinder.py]: Made this change too. It's much better. I was just afraid of creating a new method because all other methods in the file are relatively long. I really like having smaller and more methods. ^^

**Comment by aaltat** [comment]
I can take a look at this week.

**Comment by aaltat** [comment]
And I was sick last week, I will try to find time this week.

**Comment by aaltat** [comment]
By looking from mobile screen, this looks OK. I need to give second look from a bigger screen.

In mean while, could you rebase with master and create a PR which does not contain conflicts?

**Comment by aaltat** [comment]
All looks OK to me. Would you mind sorting out the conflict and it can be merged to the master?

**Comment by leods92** [comment]
@aaltat On it! Sorry it took me so long to reply. Weekend's here now! 💃

**Comment by leods92** [comment]
@aaltat Rebased and Travis is green! ✅ 😄

**Comment by aaltat** [comment]
No problem. I will try to take a look later and if all is well then merge to master.
